api-specifications.md

```diff
@@ -96,6 +104,161 @@ Create a course (anyone logged in). Used for outside Exchange.
 * Login required
 * Course already exists
 
+### /api/instructor: Course instructor management
+
+#### POST /api/instructor/&lt;course_id&gt;/&lt;instructor_id&gt;
+*Add an instructor to the course. (instructors only)*
+
+##### Request (HTTP POST data)
+```
+first_name=/*instructor name*/&
+last_name=/*instructor last name*/&
+email=/*instructor email*/
+```
+
+##### Response
+```javascript
+{
+    "success": true
+}
+```
+
+##### Error messages
+* Login required
+* Permission denied
+* Course not found
+
+#### GET /api/instructor/&lt;course_id&gt;/&lt;instructor_id&gt;
+*Gets information about a course instructor. (instructors+students)*
+
+##### Response
+```javascript
+{
+    "success": true,
+    "first_name": /*instructor name*/,
+    "last_name": /*instructor last name*/,
+    "email": /*instructor email*/
+}
+```
+
+##### Error messages
+* Login required
+* Permission denied
+* Course not found
+* Instructor not found
+
+#### DELETE /api/instructor/&lt;course_id&gt;/&lt;instructor_id&gt;
+*Remove a course instructor (instructors only)*
+
+##### Error messages
+* Login required
+* Permission denied
+* Course not found
+* Instructor not found
+
+### /api/instructors: List course instructors
+
+#### GET /api/instructors/&lt;course_id&gt;
+*Gets information about all course instructors. (instructors+students)*
+
+##### Response
+```javascript
+{
+    "success": true,
+    "instructors":
+    [
+        {
+            "first_name": /* instructor name*/,
+            "last_name": /* instructor lastname */,
+            "email": /* instructor email */,
+            "instructor_id": /* instructor ID */
+        },
+        ...
+    ]
+}
+```
+
+#### Error messages
+* Login required
+* Permission denied
+* Course not found
+
+### /api/student: Student management
+
+#### POST /api/student/&lt;course_id&gt;/&lt;student_id&gt;
+*Create or update a student. (instructors only)*
+
+##### Request (HTTP POST data)
+```
+first_name=/*student name*/&
+last_name=/*student last name*/&
+email=/*student email*/
+```
```

Comment:
This change will also change the student's record on other courses. Is this the desired behaviour?

Reply:
Yes, I believe every student should have the same name, last name, email and student_id across all of their courses. So if this information is updated in one course it should probably be updated in another course as well. Unless there are any cons to this?

Reply:
A student will be able to corrupt the record of a class he/she is taking. How is user information (first name, last name) currently stored in the JupyterHub system? I think we should make student_id = email if possible, because this is unique and cannot be changed.

Reply:
Only an instructor can update a student's information.

Reply:
Suppose we have a course "c1" with instructor "i1" and students "s1 <s1@ucdavis.edu>", "s2 <s2@ucdavis.edu>". 
Then s1 can create a course c2 with s1 as instructor and s2 as student. 
Then s1 can change s1's email to "<s2@ucdavis.edu>" and s2's email to "<s1@ucdavis.edu>", because s1 is instructor of c2, and this change also affects c1.

Reply:
I see, then changes to a student should be contained to that course only.

Reply:
There are difficulties in implementing this in backend. 
* The database structure need to be changed. 
* There may be a student with two names in two different classes. 

**How is user information (first name, last name) currently stored in the JupyterHub system?**

Reply:
I believe currently changes are made on a course by course basis, though I have to look more closely to confirm. You can see more about the current database for nbgrader in `nbgrader/nbgrader/api.py` particularly they have a function called `update_or_create_student`

Reply:
- `username` should be the JupyterHub username.
- `student_id` == `username` if the user is a student in the course.
- Names, emails, etc. should not be associated directly with the user. Every user and course combination can have different user information.
- I don't think JupyterHub stores user information. nbgrader stores the information in a database file for each course (This in in line with the previous point).

Reply:
So should we be asking for instructor information when someone is creating a course?

Reply:
Probably not because it doesn't make much sense to provide user information to initialize a course.

Reply:
Then we will have an instructor with no first name, last name, email, if the creator does not update it manually. How does it look like currently in nbgrader?